web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java

```diff
@@ -73,12 +78,28 @@ protected void doFilterInternal(HttpServletRequest request,
 				headerWriterResponse);
 
 		try {
+			if (shouldWriteHeadersEagerly) {
+				headerWriterResponse.writeHeaders();
+			}
 			filterChain.doFilter(headerWriterRequest, headerWriterResponse);
 		} finally {
-			headerWriterResponse.writeHeaders();
+			if (!shouldWriteHeadersEagerly) {
+				headerWriterResponse.writeHeaders();
+			}
 		}
 	}
 
+	/**
+	 * Allow writing headers at the beginning of the request.
+	 *
+	 * @param shouldWriteHeadersEagerly boolean to allow writing headers at the beginning of the request.
+	 * @author Ankur Pathak
+	 * @since 5.2
+	 */
+	public void setShouldWriteHeadersEagerly(boolean shouldWriteHeadersEagerly) {
```

Comment:
Let's go ahead and add a couple of tests that confirm that this functionality works. 

One test in `HeaderWriterFilterTests` that confirms that when using the setter, then the headers are written early.

And one in `HeadersConfigurerTests` that confirms that the setting can be made via an `ObjectPostProcessor`. (If you haven't done that before, you can see an example in the ticket description.) In this case, I think you can just confirm that the headers still get written.

Reply:
```
http
    .addObjectPostProcessor(new ObjectPostProcessor<HeaderWriterFilter>() {
        public HeaderW riterFilter postProcess(HeaderWriterFilter filter) {
            filter.setShouldWriteHeadersEagerly(true);
            return filter;
        }
    }
```
I can not find any addObjectPostProcessor method in HttpSecurity. Please check this url:
https://docs.spring.io/spring-security/site/docs/5.2.0.BUILD-SNAPSHOT/api/org/springframework/security/config/annotation/web/builders/HttpSecurity.html

Reply:
```@EnableWebSecurity
	static class HeadersAtTheBeginningOfRequestConfig extends WebSecurityConfigurerAdapter {
        @Override
        protected void configure(HttpSecurity http) throws Exception {
            http
                    .headers()
                    .addObjectPostProcessor(new ObjectPostProcessor<HeaderWriterFilter>() {
						@Override
						HeaderWriterFilter postProcess(HeaderWriterFilter filter) {
							filter.shouldWriteHeadersEagerly = true
							return filter
						}
					})
        }
    }
```
This inside groovy is giving me errors below:
![screenshot from 2019-02-08 09-11-54](https://user-images.githubusercontent.com/6482301/52457785-a45f5f00-2b81-11e9-881b-75a9ec5713e1.png)

Reply:
I am done with changes.

Reply:
> Let's go ahead and add a couple of tests that confirm that this functionality works.
> 
> One test in `HeaderWriterFilterTests` that confirms that when using the setter, then the headers are written early.
> 
> And one in `HeadersConfigurerTests` that confirms that the setting can be made via an `ObjectPostProcessor`. (If you haven't done that before, you can see an example in the ticket description.) In this case, I think you can just confirm that the headers still get written.

Done.

Reply:
Looks like you found it, yes, the correct way is:

```java
http
    .headers()
        .withObjectPostProcessor(...
```

Reply:
@jzheaux You can review it. Its complete.